### PR TITLE
Contributing guide

### PR DESCRIPTION
**Pull request by rajsite**
Add Vireo GitHub workflow and testing recommendation steps

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,36 +1,210 @@
 # Contributing to Vireo
 
-## Getting Started
-To begin development on Vireo, make sure to start by [forking the repository](https://help.github.com/fork-a-repo/).
+This guide will walkthrough the different workflows recommended for developing Vireo.
```
Comment (mehagar) [inline CONTRIBUTING.md]: walkthrough with no space is a noun/adjective. Use "walk through" instead. (http://www.dictionary.com/browse/walkthrough)

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,36 +1,210 @@
...
 
-### Travis-CI
+Your fork will be available online at github.com/YOUR_USER_NAME/VireoSDK
+
+## Connect fork to CI servers
+Vireo uses Travis and AppVeyor to build and test the Vireo source code automatically when changes are made.
+You are encouraged to push changes to your fork when you have changes and to make branches for experiments.
+The fork is your playground for development and experimentation.
+
+By default when you create a new fork the Travis and AppVeyor services are not enabled until you give them permission to build.
```
Comment (mehagar) [inline CONTRIBUTING.md]: "By default," (add comma)

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,36 +1,210 @@
...
+
+### Travis CI
 1. Go to [Travis-CI](https://travis-ci.org/first_sync) website and login with the github account associated with your Vireo fork
 2. (If this is your first time logging in, you can skip this step) On the left side of the page that appears click on the + sign, right next to My repositories
-3. On the page that appears, click on the repository switch to enable it
+3. On the page that appears, click on the repository switch by your fork to enable it
 
-### AppVeyor-CI
+### AppVeyor CI
 1. Go to the [AppVeyor](https://www.appveyor.com/) website and login with the github account associated with your Vireo fork
```
Comment (mehagar) [inline CONTRIBUTING.md]: "login" -> "log in" (http://grammarist.com/spelling/log-in-login/)

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,36 +1,210 @@
...
+> git remote -v
+origin  https://github.com/YOUR_USER_NAME/VireoSDK.git (fetch)
+origin  https://github.com/YOUR_USER_NAME/VireoSDK.git (push)
+upstream https://github.com/ni/VireoSDK.git (fetch)
+upstream https://github.com/ni/VireoSDK.git (push)
+```
+
+If you used GitHub Desktop to clone the fork you should see both the remotes listed above.
+
+If you used the git commandline tool to clone the fork you may need to [add the upstream remote](https://help.github.com/articles/fork-a-repo/#step-3-configure-git-to-sync-your-fork-with-the-original-spoon-knife-repository) to your local clone:
```
Comment (mehagar) [inline CONTRIBUTING.md]: "commandline" -> "command line"

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,36 +1,210 @@
...
 
-1. Make your changes
-* Run the tests (see the README for more information about testing)
-* Add any additional tests if needed to test your fix or feature
-* Submit a pull request to the proper branch (see below)
-* The [TravisCI Build](https://travis-ci.org/ni/VireoSDK/pull_requests) will pull your changes and validate that all of the tests pass before allowing the pull request to be merged.
-* If the CI build passes, your pull request can now be merged. If not, the request will not be merged until the CI passes.
+# Development workflow
+In the fork-pull workflow new development is performed in small short-lived branches in your fork and a pull request is submitted to have those changes pulled in upstream.
+Generally you should not work directly in master of your fork but instead create a branch from master for development.
```
Comment (mehagar) [inline CONTRIBUTING.md]: "Generally, you should not work directly in the master branch of your fork"

File: CONTRIBUTING.md (hunk 4 above)
Comment (spathiwa) [reply CONTRIBUTING.md]: Actually, "command-line", when used as adjective.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1,36 +1,210 @@
...
+
+When changes are pulled in to upstream from your branch they can be either merged directly or modified (squashed or rebased).
```
Comment (mehagar) [inline CONTRIBUTING.md]: "in to" -> "into" (https://www.grammarly.com/blog/into-vs-in-to/)

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -1,36 +1,210 @@
...
+The ability for pull request merges to modify commits is one reason for long-lived branches to be undesireable as maintenance is required to avoid diverging from master.
```
Comment (mehagar) [inline CONTRIBUTING.md]: "The ability of"

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -1,36 +1,210 @@
...
+Conceptually, the rebase operation will remove your current commits, bring your branch up to date with master, and then replay your commits one at a time on top of the latest changes.
+You will have the chance to resolve merge conflicts that may occur or safely cancel the rebase operation if desired.
+
+To rebase your current branch on the upstream master branch run:
+```console
+git rebase -p upstream/master
+```
+
+**NOTICE**: Be aware that performing the rebase will cause your local clone of a your branch and the fork version of your branch to diverge.
+In order to push the branch to your clone you have to rewrite the history of the branch in your clone:
```
Comment (mehagar) [inline CONTRIBUTING.md]: Did you mean "push the branch to your fork"? I thought we're referring to clone in the sentence before to mean the local branch.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -1,36 +1,210 @@
...
+**DO NOT REWRITE HISTORY IN SHARED BRANCHES**: Your fork is your playground for development and as long as you are the only developer contributing to a branch it is safe to rewrite the history of the branch.
+It is encouraged before submission to rebase your changes on top of master and to remove extraneous commits to improve the commit history.
+However, if multiple developers are doing development in a branch it requires coordination before rewriting history using rebase.
+
+# Testing Local Vireo Changes
+The [README.md](README.md) describes how to test your Vireo changes locally on your development machine with the tests in VireoSDK.
+This section will instead focus on testing your local Vireo changes in other applications that consume Vireo.
+
+## Testing vireo.js in a JavaScript application
+The vireo.js build is packaged as an npm package to make it consumeable in JavaScript applications.
```
Comment (mehagar) [inline CONTRIBUTING.md]: Consider linking to the npm website from "npm"

**Comment by gleonoliva** [comment]
Examples of expected output of commands was a great idea. Thanks!

**Pull request merged**